src/stream.rs

```diff
@@ -0,0 +1,218 @@
+                    stream,
+                ),
+            }
+        }
+
+        /// Allows writing to a borrowed stream. The lifetime of the borrow is not tracked.
+        ///
+        /// # Safety
+        ///
+        /// The value that `stream` points to must live at least until the underlying `cairo_surface_t`
+        /// (which maybe be longer then the Rust `PdfSurface` wrapper, because of reference-counting),
+        /// or until the output stream is removed from the surface with [`Surface::take_output_stream`].
+        ///
+        /// Since the former is hard to track for sure, the latter is strongly recommended.
+        /// The concrete type behind the `Box<dyn Any>` value returned by `take_output_stream`
+        /// is private, so you won’t be able to downcast it.
+        /// But removing it anyway ensures that later writes do no go through a dangling pointer.
+        pub unsafe fn for_raw_stream<W: io::Write + 'static>(width: f64, height: f64, stream: *mut W) -> Self {
+            Self {
+                inner: Surface::_for_raw_stream(
+                    ffi::$constructor_ffi,
+                    width,
+                    height,
+                    stream,
+                ),
+            }
+        }
+    };
+}
+
+impl Surface {
+    pub(crate) fn _for_stream<W: io::Write + 'static>(
+        constructor: Constructor,
+        width: f64,
+        height: f64,
+        stream: W,
+    ) -> Self {
+        let env_rc = Rc::new(UnsafeCell::new(CallbackEnvironment {
+            stream: Some(Box::new(stream)),
+            io_error: None,
+        }));
+        let env: *const UnsafeCell<CallbackEnvironment> = &*env_rc;
+        unsafe {
+            let ptr = constructor(Some(write_callback::<W>), env as *mut c_void, width, height);
+            let surface = Surface::from_raw_full(ptr);
+            surface.set_user_data(&STREAM_CALLBACK_ENVIRONMENT, env_rc);
+            surface
+        }
+    }
+
+    pub(crate) unsafe fn _for_raw_stream<W: io::Write + 'static>(
+        constructor: Constructor,
+        width: f64,
+        height: f64,
+        stream: *mut W,
+    ) -> Self {
+        Self::_for_stream(constructor, width, height, RawStream(stream))
+    }
+
+    fn with_stream_env<R, F>(&self, f: F) -> Option<R>
+        where F: FnOnce(&mut CallbackEnvironment) -> Option<R>
+    {
+        let env = self.get_user_data_ptr(&STREAM_CALLBACK_ENVIRONMENT)?;
+
+        // Safety: contract of `get_user_data_ptr`
+        let env = unsafe { env.as_ref() };
+
+        with_mut_env(env, f)
+    }
+
+    /// Remove and return the output stream, if any.
+    ///
+    /// This is relevant for surfaces created for example with [`PdfSurface::for_stream`].
+    /// Consider calling [`Surface::finish`] first,
+    /// to ensure that all writes to the stream are done.
+    ///
+    /// Use [`Box::downcast`] to recover the concrete type.
+    pub fn take_output_stream(&self) -> Option<Box<dyn Any>> {
```

Comment:
I’ve put these public methods on `Surface`, but they could instead be on `PdfSurface`, `PsSurface`, and `SvgSurface`.

Reply:
It would seem nicer to not pollute the normal `Surface` API with that but have it only on the types where it can actually exist, just like the data related functions on `ImageSurface`.

What was your reasoning for having it on the main `Surface` type? Maybe I'm missing an advantage of doing so :)